scripts/build.py

```diff
@@ -0,0 +1,243 @@
+#!/usr/bin/env python
+
+# Copyright 2020 Google
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Assuming pre-requisites are in place (please run `scripts/install_preqreqs.py` once on the system), 
+this script will build the firebase cpp sdk
+
+It does the following,
+- install platform specific cpp dependencies via vcpkg
+- cmake configure
+- cmake build
+
+Usage examples:
+# Build all targets with x64 architecture, Release config and 8 cpu cores fo
+python build.py --arch x64 --config Release -j 8
+
+# Build all targets with default options and also build unit tests
+python build.py --build_tests --arch x64
+
+# Build only firebase_app and firebase_auth with default options and also build unit tests
+python build.py --build_tests --arch x64 --targets firebase_app firebase_auth
+
+"""
+
+import subprocess
+import os
+import sys
+import argparse
+import platform
+
+
+def run_command(cmd, cwd=None, as_root_user=False, collect_output=False):
+ """Run a command.
+
+ Args:
+  cmd (list(str)): Command to run as a list object.
+          Eg: ['ls', '-l'].
+  cwd (str): Directory to execute the command from.
+  as_root_user (bool): Run command as root user with admin priveleges (supported on mac and linux).
+  collect_output (bool): Get output from the command. The string returned from this function
```

Comment:
I don't really love the return value EITHER being the return code of the subprocess OR the collected output. What about always returning the error code and the output in a tuple? (return_code, output_string)

Reply:
Definitely agree with this. I kind of don't like returning tuples because on the consumer end, we have to extra careful about the return values and unpacking these tuples. Maybe create 2 separate functions in our utility module?

Reply:
Python at least has nice syntax for unpacking tuples

x, y = func_that_returns_tuple()

Though in this case, it looks like you're not returning the return code? The return value is the output, which is possibly none (based on an arg). Returning the return code is unnecessary, since if it's anything other than 0, a CalledProcessError will occur. Thus if the function completes, it's safe to assume a return code of 0.

Reply:
Actually `subprocess.run` that I learnt from you is very convenient. "check" and "capture_output" are exactly the things we need. It also returns an object which contains the output of the command in an attribute called ".stdout". 
So yeah, this function is now just a wrapper over that.